firedrake/ensemble/ensemble_functionspace.py

```diff
@@ -0,0 +1,242 @@
+        return self._ensemble
+
+    @property
+    def comm(self):
+        """The spatial communicator from the :class:`firedrake.Ensemble` communicator.
+        """
+        return self._ensemble.comm
+
+    @property
+    def ensemble_comm(self):
+        """The ensemble communicator from the :class:`firedrake.Ensemble` communicator.
+        """
+        return self._ensemble.ensemble_comm
+
+    @property
+    def global_comm(self):
+        """The global communicator from the :class:`firedrake.Ensemble` communicator.
+        """
+        return self._ensemble.global_comm
+
+    @property
+    def local_spaces(self):
+        """The :function:`firedrake.FunctionSpace` on the local ensemble.comm.
+        """
+        return self._local_spaces
+
+    def mesh(self):
+        """The :class:`firedrake.Mesh` on the local ensemble.comm.
+        """
+        return self._mesh
+
+    @cached_property
+    def dual(self):
+        """The dual to this function space.
+        A :class:`firedrake.EnsembleDualSpace` if self is a :class:`firedrake.EnsembleFunctionSpace`, and vice-versa.
+        """
+        return EnsembleFunctionSpace(
+            [V.dual() for V in self.local_spaces], self.ensemble)
+
+    @cached_property
+    def nlocal_spaces(self):
+        """The total number of subspaces across all ensemble ranks.
+        """
+        return len(self.local_spaces)
+
+    @cached_property
+    def nglobal_spaces(self):
+        """The total number of subspaces across all ensemble ranks.
+        """
+        return self.ensemble_comm.allreduce(len(self.local_spaces))
+
+    @property
+    def nlocal_rank_dofs(self):
+        """The total number of dofs across all subspaces on the local MPI rank.
+        """
+        return self._full_local_space.node_set.size
+
+    @property
+    def nlocal_comm_dofs(self):
+        """The total number of dofs across all subspaces on the local ensemble.comm.
+        """
+        return self._full_local_space.dim()
+
+    @property
+    def nglobal_dofs(self):
+        """The total number of dofs across all subspaces on all ensemble ranks.
+        """
+        return self.ensemble_comm.allreduce(self.nlocal_rank_dofs)
+
+    def _component_indices(self, i):
+        """
+        Return the indices into the local mixed function storage
+        corresponding to the i-th local function space.
+        """
+        offset = sum(len(V) for V in self.local_spaces[:i])
+        return tuple(offset + i for i in range(len(self.local_spaces[i])))
+
+    def create_vec(self):
+        """Return a PETSc Vec on the global_comm with the same layout as a :class:`firedrake.EnsembleFunction` or :class:`firedrake.EnsembleCofunction` in this function space.
+        """
+        vec = PETSc.Vec().create(comm=self.global_comm)
+        vec.setSizes((self.nlocal_dofs, self.nglobal_dofs))
+        vec.setUp()
+        return vec
+
+    def local_to_global_ises(self, i):
+        # COMM_SELF or global_comm?
+        pass
```

Comment:
Clearly a TODO

Reply:
Removed. May be useful in the future but currently unneeded.